R/estimateDivergence.R

```diff
@@ -0,0 +1,152 @@
+#' @export
+#'
+#' @author Leo Lahti and Tuomas Borman. Contact: \url{microbiome.github.io}
+#' 
+#' @examples
+#' data(GlobalPatterns)
+#' tse <- GlobalPatterns
+#' 
+#' # By default, reference is median of all samples. The name of column where results
+#' # is "divergence" by default, but it can be specified. 
+#' tse <- estimateDivergence(tse)
+#' 
+#' # The method that are used to calculate distance in divergence and 
+#' # reference can be specified. Here, euclidean distance and dist function from 
+#' # stats package are used. Reference is the first sample.
+#' tse <- estimateDivergence(tse, name = "divergence_first_sample", 
+#'                           reference = assays(tse)$counts[,1], 
+#'                           FUN = stats::dist, method = "euclidean")
+#' 
+#' # Reference can also be median or mean of all samples. 
+#' # By default, divergence is calculated by using median. Here, mean is used.
+#' tse <- estimateDivergence(tse, name = "divergence_average", reference = "mean")
+#' 
+#' # All three divergence results are stored in colData.
+#' colData(tse)
+#' 
+NULL
+
+#' @rdname estimateDivergence
+#' @export
+setGeneric("estimateDivergence",signature = c("x"),
+           function(x, abund_values = "counts", name = "divergence", 
+                    reference = "median", FUN = vegan::vegdist, method = "bray", 
+                    ...)
+             standardGeneric("estimateDivergence"))
+
+#' @rdname estimateDivergence
+#' @export
+setMethod("estimateDivergence", signature = c(x="SummarizedExperiment"),
+    function(x, abund_values = "counts", name = "divergence", 
+             reference = "median", FUN = vegan::vegdist, method = "bray", 
+             ..., BPPARAM = SerialParam()){
+        
+        ################### Input check ###############
+        # Check abund_values
+        .check_assay_present(abund_values, x)
+        # Check name
+        if(!.is_non_empty_character(name) || length(name) != 1L){
+            stop("'name' must be a non-empty character value.",
+                 call. = FALSE)
+        }
+        # Check regetence
+        # If "reference" is not right: 
+        # it is null, its length does not equal to number of samples and it is not numeric,
+        # reference is not "median" or "mean"
+        reference_stop_msg <- 
+            paste0("'reference' must be a numeric vector that has lenght equal",
+                   " to number of features, or 'reference' must be either",
+                   " 'median' or 'mean'.")
+        if( is.null(reference) ){
+            stop(reference_stop_msg, call. = FALSE)
+        } else {
+            if(is.numeric(reference) && length(reference) != nrow(x)){
+                stop(reference_stop_msg, call. = FALSE)
+            }
+            if(is.character(reference) && 
+               (length(reference) != 1L || !(reference %in% c("median","mean")))){
+                stop(reference_stop_msg, call. = FALSE)
+            }
+        }
```

Comment:
I simplified this a bit. Please check, that it still works as intended by you.

Reply:
I made some changes because boolean values went through without an error. Now it should allow only two type types of input: 

- numeric and length of `nrow`
- "median" or "mean"